File: AOJ/0181/book_shelf.c
#include<stdio.h>

int m, n;
int w[100];

int isStorable(int width) {
  int row = 0, c = 0, i = 0;
  while ( i < n ) {
    if ( c + w[i] <= width ) {
      c += w[i];
      ++i;
    }
    else if ( ++row < m ) c = 0;
    else return 0;
  }
  return 1;
}

int main() {
  int i, minw, maxw, mid;

  while ( scanf("%d %d", &m, &n) == 2 && m && n ) {

    for ( i = 0; i < n; ++i ) scanf("%d", &w[i]);

    minw = 1;
    maxw = 1500000;
    if ( !isStorable(maxw) ) printf("NA\n");

    while ( maxw > minw ) {
      mid = (maxw + minw) / 2;
      if ( isStorable(mid) ) maxw = mid;
      else                   minw = mid + 1;
    }

    printf("%d\n", minw);
  }
  return 0;
}

File: AOJ/0181/book_shelf.py
def isStorable( m, n, w, width ):
    row = 1
    c = 0
    i = 0
    while i < n :
        if c + w[i] <= width :
            c += w[i]
            i += 1
        elif row < m :
            c = 0
            row += 1
        else : return False
    return True

while ( True ):
    m,n = map(int, input().split())
    if m == 0 and n == 0 : break
    w = []
    for i in range(n) : w.append(int(input()))

    minw = 1
    maxw = 1500000
    if not isStorable(m, n, w, maxw) : print("NA")
    
    while maxw > minw :
        mid = int((maxw + minw) / 2)
        if isStorable(m, n, w, mid) : maxw = mid
        else : minw = mid + 1
    print(minw)
